pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -241,9 +241,16 @@ func (cb *ClusterBuilder) buildDefaultCluster(name string, discoveryType cluster
 	return c
 }
 
-func (cb *ClusterBuilder) buildInboundClusterForPortOrUDS(proxy *model.Proxy, instance *model.ServiceInstance, bind string) *cluster.Cluster {
+// buildInboundClusterForPortOrUDS constructs a single inbound listener. The cluster will be bound to
+// `inbound|clusterPort||`, and send traffic to <bind>:<instance.Endpoint.EndpointPort>. A workload
+// will have a single inbound cluster per port. In general this works properly, with the exception of
+// the Service-oriented DestinationRule, and upstream protocol selection. Our documentation currently
+// requires a single protocol per port, and the DestinationRule issue is slated to move to Sidecar.
+// Note: clusterPort and instance.Endpoint.EndpointPort are identical for standard Services; however,
+// Sidecar.Ingress allows these to be different.
+func (cb *ClusterBuilder) buildInboundClusterForPortOrUDS(proxy *model.Proxy, clusterPort int, instance *model.ServiceInstance, bind string) *cluster.Cluster {
 	clusterName := util.BuildInboundSubsetKey(proxy, instance.ServicePort.Name,
-		instance.Service.Hostname, instance.ServicePort.Port)
+		instance.Service.Hostname, instance.ServicePort.Port, clusterPort)
 	localityLbEndpoints := buildInboundLocalityLbEndpoints(bind, instance.Endpoint.EndpointPort)
 	localCluster := cb.buildDefaultCluster(clusterName, cluster.Cluster_STATIC, localityLbEndpoints,
```

Comment:
This function, which builds the telemetry metadata, does not take the cluster port as an argument. So I guess the cluster does not have proper service metadata attached.

Reply:
Aren't there tests for this? I don't see why it needs the cluster port, it finds the instances from `opts.proxy.ServiceInstances`

Reply:
If we have two services:

foo: 80 -> 8081
bar: 80 -> 8082

This call will be called for cluster 'inbound||8081' and 'inbound|8082'. For both of them, it will be called with port "80", look through all `opts.proxy.ServiceInstances`, and find `foo, bar`

Reply:
Yeah, we don't have a test covers overlapping services I think. It would be nice to add a unit test for this case with cluster metadata build.

Reply:
This should be fixed now, improved the tests as well

Reply:
Was the original naming change recent ?
Can we send both names and then remove the old name at some later point ?
This is the kind of this control plane should be able to do, especially when it is self inflicted.

Reply:
1.8.1: `inbound|<workload>||`
1.8.0: `inbound|<service>||`
older: `inbound|<service>|<port-name>|<hostname>`

> Can we send both names and then remove the old name at some later point ?

It is tricky since the old and new format overlap if we have a matching service port and target port. This may lead to a short period of time where we send traffic to the wrong service (since CDS+LDS is not synchronized).

The main alternative I can think of is to completely change the format, maybe something like `inbound|<workload port>||inbound`.

This would also lead to 2x clusters though. The current code is instead having 3 branches to determine the format based on version which sounds horribly complex but in reality doesn't seem to be too bad. We don't need to support it for long since we (at least historically) haven't supported 1.x.0 to 1.y.0, just some 1.x.z -> 1.y.0 as we may need to make compatibility changes like this. We could also continue support into 1.9, and drop it in 1.10 without issues.

Reply:
Yeah.  Looks like sending both and dropping at a later time seems fine - but would increase the number of clusters. I think let us move forward with the current fix and if it creates problems - may be we can take the hit of sending both?